### PR TITLE
Add completions for fish

### PR DESCRIPTION
**Pull request by delphinus**
Fish does not have the comprehensive subshell feature like bash (as far as I know), so this hack creates a file containing a function declaration and will be executed in background.

I have implemented all features in the bash completion. I'm thinking to implement more features for this.

* [x] Show a description for each subcommand

  [asciicast](https://asciinema.org/a/369825)

* [x] Complete filenames in `$_nb_dir`

  [asciicast](https://asciinema.org/a/370585)

* [x] Complete filenames for `nb import` (#59)
  - In fact, I added filenames for all subcommands.

### CODE REVIEW

**Comment by 100ideas** [comment]
awesome, thanks for working on this. was just looking into how to maybe autogenerate fish completions from those provided in this repo.

**Comment by 100ideas** [comment]
Also, @delphinus I am not sure how your installed your `nb-completions.fish` file, but fish didn't recognize mine until I renamed it to `nb.fish`. I decided to store it in  `~/.config/fish/completions/nb.fish` See [fish docs on completions](http://fishshell.com/docs/current/index.html#where-to-put-completions).

completions that ship with fish all use the `<cmdname>.fish` naming pattern - see in ```ls /usr/local/share/fish/completions``` (osx)

**Comment by delphinus** [comment]
@100ideas Ah, good point! I myself also wonder why fish does not detect this ;) I've renamed to `nb.fish` and fish does good.

**Comment by delphinus** [comment]
I added description for each command completion (See [example][]). It extracts each description by parsing `nb help <command>`. This is a bit hacky and needs time because it calls `nb help` several times, I think it is in the permissible range.

[example]: https://asciinema.org/a/369825

**Comment by xwmx** [comment]
Looks super cool. We can also probably save the help text in `${NB_DIR}/.cache` to avoid going through the program. Untested example:

```bash
local _help_cache_dir="${NB_DIR}/.cache/help"
local _subcommand_help=
local _subcommand_name="${1:-}"

if [[ -f "${_help_cache_dir}/${_subcommand_name}.nb-cache" ]]
then
  _subcommand_help=$(<"${_help_cache_dir}/${_subcommand_name}.nb-cache")
else
  rm -r "${NB_DIR:?}/.cache/help"

  for __subcommand in $(nb subcommands)
  do
    printf "%s\\n" "$(nb help "${__subcommand}")" > "${_help_cache_dir}/${__subcommand}.nb-cache"
  done
fi
```

**Comment by delphinus** [comment]
I added completions for files in `$_nb_dir` ([example](https://asciinema.org/a/370585)). This makes it show all subcommands and files at once, so subcommands have disappeared below files.

```sh
> nb <Tab>
20200101123456.md
20200101234567.md
...
...
# Terminal will be filled with filenames and subcommands are not shown.
```

Hmm... this cannot be helped, I think.

When input a notbook name and complete, it shows subcommands, just to be sure.

```sh
> nb home:<Tab>
home:a                                   (Create a new note)
home:add                                 (Create a new note)
home:b            (Create, view, search, edit, and delete……)
home:bookmark     (Create, view, search, edit, and delete……)
…and 44 more rows
```

**Comment by delphinus** [comment]
@xwmx Caching is helpful for performance in completion, but cached files will not be updated when we update `nb` itself.

The biggest problem is that `_nb_subcommands` calls `nb help foobar` many times. So I'm thinking to add a command to show all help docs at once to parse description for all subcommands.

**Comment by delphinus** [comment]
I added a new option: `nb subcommands describe --all` that can show all descriptions for all commands at once. This outputs with [ASCII Delimited Text][] -- that uses joining columns with `\037` (US: Unit Separator) and joining records with `\036` (RS: Record Separator).

[ASCII Delimited Text]: https://ronaldduncan.wordpress.com/2009/10/31/text-file-formats-ascii-delimited-text-not-csv-or-tab-delimited-text/

```sh
> nb subcommands describe --all

# command and description separated by \037
help\037Usage:
  nb help [<subcommand>] [-p | --print]
  nb help [-c | --colors] | [-r | --readme] | [-s | --short] [-p | --print]

...

# the next line has the end of description, \036 (Record Separator),
# and the next record.
Shortcut Alias: `h`\036h\037Usage:
  nb help [<subcommand>] [-p | --print]
  nb help [-c | --colors] | [-r | --readme] | [-s | --short] [-p | --print]

...
```

With this change, it much improves the speed to complete at the first time.

```sh
# the 1st time to complete with creating the cache

# before
> time fish -c 'source etc/nb.fish && _nb_subcommands' > /dev/null

________________________________________________________
Executed in    9.88 secs   fish           external
   usr time    4.10 secs  175.00 micros    4.10 secs
   sys time    5.73 secs  1156.00 micros    5.73 secs

# after
> time fish -c 'source etc/nb.fish && _nb_subcommands' > /dev/null

________________________________________________________
Executed in    2.27 secs   fish           external
   usr time    1.47 secs  101.77 millis    1.36 secs
   sys time    0.93 secs   10.68 millis    0.92 secs
```

**Comment by delphinus** [comment]
```
not ok 785 'plugins' lists plugins.
# (in test file test/plugins.bats, line 20)
#   `[[ "${lines[0]}" =~ example.nb-plugin   ]]' failed
# ${status}: '0'
# ${output}: 'turquoise.nb-theme
# example.nb-plugin'
```

hmm... only Ubuntu 20.04 fails on this test. Why?

**Comment by xwmx** [comment]
@delphinus Not sure. I just updated `plugins` so output is always sorted, so it should pass again now.

**Comment by delphinus** [comment]
All done for this. Pls review @xwmx

**Comment by xwmx** [comment]
Awesome! I want to take a focused look at the completions as soon as folder support is implemented in all subcommands. There are only a few remaining. I’m hoping that folder completion will be as smooth and useful as the notebook and subcommand completions, so I want to take a close look at how that works in the different shells. I’ll be able to install this and take a close look then.

Also, is it easy to include this in `nb completions install` and should we include it there?

**Comment by delphinus** [comment]
@xwmx I updated to enable to install/check fish completions.

File: etc/nb-completion.fish [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+  end
+
+  # list all notes to complete
+  nb run ls | string replace -r '$' \tfile
+
+  if test -e $_cache_path
+    tail -n+3 $_cache_path
+
+    # write the func itself into a temporary file and execute it in background
+    set _tmp_file (mktemp -t nb-completion)
```
Comment (nbCloud91) [inline etc/nb-completion.fish:160]: This throws errors "Too few Xs" . Change to "nb-completion.XXX"

Comment (delphinus) [reply etc/nb-completion.fish]: I found `mktemp` in Linux needs `X`s that one in macOS dose not need. Fixed on 30207a4.

**Comment by jupart** [comment]
As a `fish` and `nb` user, is there anything I can do to help with this MR?

**Comment by xwmx** [comment]
I have a lot of thoughts on this. Too many to articulate easily. Overall, I think of the completions in the following priority:

1. Subcommands
2. Notebook names
3. Folder names
4. Maybe filenames
5. Everything else

Currently, Bash and Zsh completions cover 1 and 2, and I view folder name completions as the next step for those. My current view is that filename completions in all shells should be disabled by default and enabled via an optional environment variable. I’d have to experiment.

Starting with just subcommand and notebook name completions like the bash and zsh completions would make it easier to merge.

**Comment by xwmx** [comment]
I took a closer look. There have been some changes that impact this since the pull request was started. I'll work on merging the branch.

**Comment by xwmx** [comment]
Thanks for your significant patience with this. I was finally able to merge it. I pared it back so it matches the bash and zsh completion functionality. The `describe` function works differently now, we have to reimplement the `describe —all` functionality to get it to work again. I appreciate that you put it together, it looks interesting, and I would like to follow up on it. I just want to get all the completions at parity and available with the core functionality.

**Pull request merged**